desktop/cmp/treemap/TreeMap.js

```diff
@@ -0,0 +1,297 @@
+                }
+            },
+            series: [{
+                data,
+                type: 'treemap',
+                animation: false,
+                layoutAlgorithm: algorithm,
+                borderWidth: 0,
+                dataLabels: {
+                    enabled: true,
+                    allowOverlap: false,
+                    align: 'left',
+                    verticalAlign: 'top',
+                    style: {textOutline: 'none', visibility: 'hidden'}
+                },
+                events: {click: this.onClick}
+            }]
+        };
+    }
+
+    //----------------------
+    // Click handling
+    //----------------------
+    onClick = (e) => {
+        this._clickCount++;
+        this._debouncedClickHandler(e.point.record, e);
+        if (this._clickCount >= 2) this._debouncedClickHandler.flush();
+    };
+
+    clickHandler(record, e) {
+        const {onClick, onDoubleClick} = this.model;
+        if (onClick && this._clickCount === 1) {
+            onClick(record, e);
+        } else if (onDoubleClick && this._clickCount === 2) {
+            onDoubleClick(record, e);
+        }
+        this._clickCount = 0;
+    }
+
+    //----------------------
+    // Selection handling
+    //----------------------
+    syncSelection() {
+        if (!this._chart) return;
+
+        const {selectedIds, maxDepth, gridModel} = this.model;
+        let toSelect = [...selectedIds];
+
+        // Fallback to parent node if selection exceeds max depth
+        if (maxDepth && gridModel && gridModel.treeMode) {
+            toSelect = selectedIds.map(id => {
+                const delim = Cube.RECORD_ID_DELIMITER,
+                    parts = id.split(delim),
+                    idDepth = parts.filter(p => p !== 'root').length,
+                    toRemove = idDepth - maxDepth;
+
+                return toRemove < 1 ? id : parts.slice(0, -toRemove).join(delim);
+            });
+        }
+
+        // Update selection in chart
+        this._chart.series[0].data.forEach(node => {
+            node.select(toSelect.includes(node.id), true);
+        });
+    }
+
+    //----------------------
+    // Labels
+    //----------------------
+    async updateLabelVisibilityAsync() {
+        if (!this._chart) return;
+
+        // Show / hide labels by comparing label size to node size
+        await forEachAsync(this._chart.series[0].data, node => {
+            if (node.dataLabel) {
+                const buffer = 10,
+                    tooSmallWidth = (node.dataLabel.width + buffer) > node.graphic.element.width.baseVal.value,
+                    tooSmallHeight = (node.dataLabel.height + buffer) > node.graphic.element.height.baseVal.value,
+                    style = tooSmallWidth || tooSmallHeight ? {visibility: 'hidden'} : {visibility: 'visible'};
+
+                node.update({dataLabels: {style}}, false, false);
+            }
+        });
+
+        this._chart.redraw();
```

Comment:
Would like to profile this to understand the price we're paying for the (admittedly very nice, probably essential) behavior here.  Could there be a way to do with CSS?